src/wayland/shell/xdg/handlers/surface/popup.rs

```diff
@@ -1,38 +1,61 @@
 use std::sync::atomic::Ordering;
 
 use crate::{
-    utils::Serial,
+    input::{Seat, SeatHandler},
+    utils::{Serial, SERIAL_COUNTER},
     wayland::{
-        compositor,
+        compositor, seat,
         shell::xdg::{SurfaceCachedState, XdgPopupSurfaceData, XdgPositionerUserData},
     },
 };
 
 use wayland_protocols::xdg::shell::server::xdg_popup::{self, XdgPopup};
 
-use wayland_server::{backend::ClientId, DataInit, Dispatch, DisplayHandle, Resource};
+use wayland_server::{
+    backend::ClientId, protocol::wl_seat::WlSeat, DataInit, Dispatch, DisplayHandle, Resource,
+};
 
 use super::{PopupConfigure, XdgShellHandler, XdgShellState, XdgShellSurfaceUserData, XdgSurfaceUserData};
 
 impl<D> Dispatch<XdgPopup, XdgShellSurfaceUserData, D> for XdgShellState
 where
     D: Dispatch<XdgPopup, XdgShellSurfaceUserData>,
     D: XdgShellHandler,
+    D: SeatHandler,
     D: 'static,
 {
     fn request(
         state: &mut D,
-        _client: &wayland_server::Client,
+        client: &wayland_server::Client,
         popup: &XdgPopup,
         request: xdg_popup::Request,
         data: &XdgShellSurfaceUserData,
-        _dh: &DisplayHandle,
+        dh: &DisplayHandle,
         _data_init: &mut DataInit<'_, D>,
     ) {
         match request {
             xdg_popup::Request::Destroy => {
                 if let Some(surface_data) = data.xdg_surface.data::<XdgSurfaceUserData>() {
                     surface_data.has_active_role.store(false, Ordering::Release);
+
+                    // If a pointer has focus (entered the popup's surface), protocol states that a
+                    // leave event must be generated before the next enter.
+                    let mut ids = vec![];
+                    let _ = dh.backend_handle().with_all_objects_for(client.id(), |id| {
+                        ids.push(id);
+                    });
+                    for id in ids {
+                        let Ok(seat) = WlSeat::from_id(dh, id) else {
+                            continue;
+                        };
+                        let seat: Seat<D> = Seat::from_resource(&seat).unwrap();
```

Comment:
*Untested*, but I think we can simplify this by doing something like this:

```suggestion

                    // If a pointer has focus (entered the popup's surface), protocol states that a
                    // leave event must be generated before the next enter.
                    let seat_state = state.seat_state();
                    for seat in &seat_state.seats {
```

Other than that, looks fine to me.